packages/core/src/lib/ConnectionManager.ts

```diff
@@ -0,0 +1,336 @@
+    );
+    this.attachEventListeners(options.pingKeepAlive, options.relayKeepAlive);
+  }
+
+  public startDiscoveryConnectionService(
+    maxBootstrapPeersAllowed = DEFAULT_MAX_BOOTSTRAP_PEERS_ALLOWED,
+    interval = DEFAULT_PEER_DISCOVERY_CONNECTION_INTERVAL
+  ): void {
+    if (this.isConnectionServiceStarted) {
+      throw new Error("Connection service already started");
+    }
+
+    this.isConnectionServiceStarted = true;
+
+    this.dialPeersInInterval(maxBootstrapPeersAllowed, interval);
+  }
+
+  public stopAllKeepAlives(): void {
+    for (const timer of [
+      ...Object.values(this.pingKeepAliveTimers),
+      ...Object.values(this.relayKeepAliveTimers),
+    ]) {
+      clearInterval(timer);
+    }
+
+    this.pingKeepAliveTimers = {};
+    this.relayKeepAliveTimers = {};
+  }
+
+  private attachEventListeners(
+    pingKeepAlive: number,
+    relayKeepAlive: number
+  ): void {
+    this.libp2p.connectionManager.addEventListener("peer:connect", (evt) => {
+      this.startKeepAlive(evt.detail.remotePeer, pingKeepAlive, relayKeepAlive);
+    });
+
+    /**
+     * NOTE: Event is not being emitted on closing nor losing a connection.
+     * @see https://github.com/libp2p/js-libp2p/issues/939
+     * @see https://github.com/status-im/js-waku/issues/252
+     *
+     * >This event will be triggered anytime we are disconnected from another peer,
+     * >regardless of the circumstances of that disconnection.
+     * >If we happen to have multiple connections to a peer,
+     * >this event will **only** be triggered when the last connection is closed.
+     * @see https://github.com/libp2p/js-libp2p/blob/bad9e8c0ff58d60a78314077720c82ae331cc55b/doc/API.md?plain=1#L2100
+     */
+    this.libp2p.connectionManager.addEventListener("peer:disconnect", (evt) => {
+      this.stopKeepAlive(evt.detail.remotePeer);
+    });
+  }
+
+  private dialPeersInInterval(
+    maxBootstrapPeersAllowed: number,
+    interval: number
+  ): void {
+    // increase interval after every attempt
+    // -> 1st interval: 10 seconds x 2 = 20 seconds
+    // -> 2nd interval: 10 seconds x 3 = 40 seconds
+    // and so on
+    const newInterval = this.dialAttempt * interval;
+    log(`Dialing next set of discovered peers in ${newInterval} ms`);
+
+    // recursively run this function with increase in time
+    this.dialDiscoveredPeers(maxBootstrapPeersAllowed)
+      .then(() => {
+        setTimeout(() => {
+          this.dialPeersInInterval(maxBootstrapPeersAllowed, interval);
+        }, newInterval);
+        this.dialAttempt++;
+      })
+      .catch((err) => {
+        log("Error dialing discovered peers", err);
+      });
+  }
+
+  private async dialDiscoveredPeers(
+    maxBootstrapPeersAllowed: number
+  ): Promise<void> {
+    const {
+      dialableBootstrapPeers,
+      dialableNonBootstrapPeers,
+      bootstrapConnections,
+    } = await this.fetchUpdatedState();
+
+    // find & dial peers found via `dns-discovery`
```

Comment:
```suggestion
    // find & dial peers found via `dns-discovery` or other bootstrap discovery method
```

Reply:
I propose to write:

> peers found via `dns-discovery` or other **bootstrap** discovery method

You say

> not true. We Are only looking at **bootstrap** peers

I fail to see how our two statements differ.